File: csharp/StoryletFramework/Storylets.cs
/*
 * This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
 * Copyright (c) 2025 Ian Thomas
 */

using ExpressionParser;

namespace StoryletFramework;

using Context = Dictionary<string, object>;
using KeyMap = Dictionary<string, object>;

public class Storylet
{
    private static readonly Parser expressionParser = new Parser();

    public string Id { get; private set; }
    public dynamic? Content { get; set; }

    // Redraw settings
    public const int REDRAW_ALWAYS = 0;
    public const int REDRAW_NEVER = -1;
    public int Redraw { get; set; } = REDRAW_ALWAYS;

    // Updates to context when played
    public KeyMap? Outcomes { get; set; }

    // Precompiled condition and priority
    private ExpressionNode? _condition;
    private object _priority = 0;

    // The next draw this should be available
    private int _nextPlay = 0;

    internal Deck? deck = null;

    public Storylet(string id)
    {
        Id = id;
    }

    // Reset the redraw counter
    public void Reset()
    {
        _nextPlay = 0;
    }

    // Set condition as a precompiled expression
    public void SetCondition(string? text)
    {
        _condition = null;
        if (!string.IsNullOrEmpty(text))
        {
            _condition = expressionParser.Parse(text);
        }
    }

    // Evaluate condition using the current context. Returns true if no condition is set.
    public bool CheckCondition(Context context, List<string>? dumpEval = null)
    {
        if (_condition == null)
            return true;

        dumpEval?.Add($"Evaluating condition for {Id}");
        return Convert.ToBoolean(_condition.Evaluate(context, dumpEval));
    }

    // Set priority to a fixed number or a precompiled expression
    public void SetPriority(object numOrExpression)
    {
        if (numOrExpression is int || numOrExpression is double)
        {
            _priority = numOrExpression;
        }
        else if (numOrExpression is string expression)
        {
            _priority = expressionParser.Parse(expression);
        }
        else
        {
            throw new ArgumentException("Priority must be a number or a valid expression.");
        }
    }

    // Evaluate priority using the current context
    public int CalcCurrentPriority(Context context, bool useSpecificity = true, List<string>? dumpEval = null)
    {
        int workingPriority = 0;

        if (_priority is int || _priority is double)
        {
            workingPriority = Convert.ToInt32(_priority);
        }
        else if (_priority is ExpressionNode expression)
        {
            dumpEval?.Add($"Evaluating priority for {Id}");
            workingPriority = Convert.ToInt32(expression.Evaluate(context, dumpEval));
        }

        if (useSpecificity)
        {
            workingPriority *= 100;
            if (_condition != null)
            {
                workingPriority += _condition.Specificity;
            }
        }

        return workingPriority;
    }

    // Check if the storylet is available to draw based on redraw rules
    public bool CanDraw(int currentPlay)
    {
        if (Redraw == REDRAW_NEVER && _nextPlay < 0)
            return false;
        if (Redraw == REDRAW_ALWAYS)
            return true;
        return currentPlay >= _nextPlay;
    }

    // Update the redraw counter when the storylet is drawn
    internal void OnPlayed(int currentDraw, Context context, string outcome="default", List<string>? dumpEval = null)
    {
        if (dumpEval != null)
        {
            dumpEval.Add($"Storylet {Id} played with outcome '{outcome}' at draw {currentDraw}.");
        }

        if (Redraw == REDRAW_NEVER)
        {
            _nextPlay = -1;
        }
        else
        {
            _nextPlay = currentDraw + Redraw;
        }

        if (Outcomes != null && Outcomes.ContainsKey(outcome))
        {
            var updates = Outcomes[outcome];
            if (dumpEval != null) {
                dumpEval.Add($"Updating context for {Id} with outcome '${outcome}");
            }
            ContextUtils.UpdateContext(context, (KeyMap)updates, dumpEval);
        }
    }

      // Convenience. Play the storylet. Updates draw counters and applies any updateOnPlayed
    public void Play(string outcome = "default", List<string>? dumpEval = null) {
        if (deck == null) {
            throw new Exception("Storylet not in deck.");
        }
        deck.Play(this, outcome, dumpEval);
    }

}

public class Deck
{
    public bool UseSpecificity { get; set; } = false;
    private readonly Dictionary<string, Storylet> _all = new();
    private int _currentPlay = 0;
    public readonly Context Context;

    public Deck(Context? context = null)
    {
        Context = context ?? new Context();
    }

    // Reset the whole deck, including all redraw counters
    public void Reset()
    {
        _currentPlay = 0;
        foreach (var storylet in _all.Values)
        {
            storylet.Reset();
        }
    }

    // Reshuffle and draw from the deck
    public List<Storylet> Draw(int count = -1, Func<Storylet, bool>? filter = null, List<string>? dumpEval = null)
    {
        var priorityMap = new SortedDictionary<int, List<Storylet>>(Comparer<int>.Create((a, b) => b.CompareTo(a)));
        var toProcess = _all.Values.ToList();

        while (toProcess.Count > 0)  
        {
            var storylet = toProcess[0];
            toProcess.RemoveAt(0);

            if (!storylet.CanDraw(_currentPlay))
                continue;

            if (filter != null && !filter(storylet))
                continue;

            if (!storylet.CheckCondition(Context, dumpEval))
                continue;

            var priority = storylet.CalcCurrentPriority(Context, UseSpecificity, dumpEval);

            if (!priorityMap.ContainsKey(priority))
            {
                priorityMap[priority] = new List<Storylet>();
            }

            priorityMap[priority].Add(storylet);
        }

        var drawPile = new List<Storylet>();

        foreach (var bucket in priorityMap.Values)
        {
            Utils.ShuffleArray(bucket);
            drawPile.AddRange(bucket);
            if (count>-1 && drawPile.Count >= count)
            {
                break;
            }
        }

        return count > -1 ? [.. drawPile.Take(count)] : drawPile;
    }

    public List<Storylet> DrawAndPlay(int count = -1, Func<Storylet, bool>? filter = null, string outcome="default", List<string>? dumpEval = null)
    {
        var drawPile = Draw(count, filter, dumpEval);
        foreach (var storylet in drawPile)
        {
            Play(storylet, outcome, dumpEval);
        }
        return drawPile;
    }

    public Storylet? DrawSingle(Func<Storylet, bool>? filter = null, List<string>? dumpEval = null)
    {
        var drawPile = Draw(1, filter, dumpEval);
        if (drawPile.Count > 0)
        {
            return drawPile[0];
        }
        return null;
    }

    public Storylet? DrawAndPlaySingle(Func<Storylet, bool>? filter = null, string outcome="default", List<string>? dumpEval = null)
    {
        var storylet = DrawSingle(filter, dumpEval);
        if (storylet != null)
        {
            Play(storylet, outcome, dumpEval);
        }
        return storylet;
    }

    public Storylet? GetStorylet(string id)
    {
        if (_all.TryGetValue(id, out var storylet))
        {
            return storylet;
        }
        return null;
    }

    public void AddStorylet(Storylet storylet)
    {
        if (_all.ContainsKey(storylet.Id))
            throw new ArgumentException($"Duplicate storylet id: '{storylet.Id}'.");

        _all[storylet.Id] = storylet;
        storylet.deck = this;
    }

    public void Play(Storylet storylet, string outcome = "default", List<string>? dumpEval = null) {
        _currentPlay++;
        storylet.OnPlayed(_currentPlay, Context, outcome, dumpEval);
    }
}

File: python/src/storylet_framework/storylets.py
# This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
# Copyright (c) 2025 Ian Thomas

import sys
import os
from .utils import shuffle_array
from .context import update_context

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../lib/expression-parser")))

from expression_parser.parser import Parser
expression_parser = Parser()

REDRAW_ALWAYS = 0
REDRAW_NEVER = -1

class Storylet:
    def __init__(self, id):
        # Unique ID of the storylet
        self.id = id

        # Application-defined content
        self.content = {}

        # Redraw setting:
        #   >0 = how many draws before this is available again,
        #   REDRAW_ALWAYS = always available,
        #   REDRAW_NEVER = a one-shot
        self.redraw = REDRAW_ALWAYS

        # Updates to context, processed when this storylet is drawn
        self.update_on_drawn = None

        # Precompiled
        self._condition = None
        # Might be absolute value, might be an expression
        self._priority = 0
        # The next draw this should be available
        self._next_draw = 0

    # Reset the redraw counter
    def reset(self):
        self._next_draw = 0

    # Set text of expression. It'll be precompiled
    @property
    def condition(self):
        return self._condition

    @condition.setter
    def condition(self, text):
        self._condition = None
        if text:
            self._condition = expression_parser.parse(text)

    # Evaluate condition using current context. Return boolean. If no condition, returns True
    def check_condition(self, context, dump_eval=None):
        if not self._condition:
            return True
        if dump_eval is not None:
            dump_eval.append(f"Evaluating condition for {self.id}")
        return self._condition.evaluate(context, dump_eval)

    # Set priority to a fixed number, or to an expression which will be precompiled
    @property
    def priority(self):
        return self._priority

    @priority.setter
    def priority(self, num_or_expression):
        if isinstance(num_or_expression, (int, float)):
            self._priority = num_or_expression
        else:
            self._priority = expression_parser.parse(num_or_expression)

    # Evaluate priority using current context
    # If use_specificity is True, will produce a higher number if the storylet's expression is more complex
    def calc_current_priority(self, context, use_specificity=True, dump_eval=None):
        working_priority = 0
        if isinstance(self._priority, (int, float)):
            working_priority = self._priority
        else:
            if dump_eval is not None:
                dump_eval.append(f"Evaluating priority for {self.id}")
            working_priority = self._priority.evaluate(context, dump_eval)

        if use_specificity:
            working_priority *= 100
            if self._condition is not None:
                working_priority += self._condition.specificity

        return working_priority

    # True if the card is available to draw according to its redraw rules
    def can_draw(self, current_draw):
        if self.redraw == REDRAW_NEVER and self._next_draw < 0:
            return False
        if self.redraw == REDRAW_ALWAYS:
            return True
        return current_draw >= self._next_draw

    # Call when actually drawn - updates the redraw counter
    def drawn(self, current_draw):
        if self.redraw == REDRAW_NEVER:
            self._next_draw = -1
        else:
            self._next_draw = current_draw + self.redraw
    

class Deck:
    def __init__(self, context={}):

        # If true, storylet.priority is still used as the base, but
        # within priorities, more complex conditions (thus more specific) are
        # treated as higher priority
        self.use_specificity = False

        # How many storylets to process on each call to update(), if using async reshuffles
        self.async_reshuffle_count = 10

        # Complete set of storylets
        self._all = {}

        # Currently calculated pile
        self._draw_pile = []

        # Keeps a count of the number of draws. Used to keep redraw rules correct.
        # Call reset() to start from scratch.
        self._current_draw = 0

        # Context to be used for all expression evaluations.
        self.context = context

        # Used to deal with in-progress reshuffles.
        self._reshuffle_state = {
            "callback": None,
            "to_process": [],
            "filter": None,
            "priority_map": {},
            "dump_eval": None,
        }

    def reset(self):
        self._current_draw = 0
        for storylet in self._all.values():
            storylet.reset()

    def reshuffle(self, filter=None, dump_eval=None):
        if self.async_reshuffle_in_progress():
            raise RuntimeError("Async reshuffle in progress, can't call reshuffle()")

        self._reshuffle_prep(filter, dump_eval)
        # Reshuffle everything at once
        self._reshuffle_do_chunk(len(self._reshuffle_state["to_process"]))
        self._reshuffle_finalize()

    def reshuffle_async(self, callback, filter=None, dump_eval=None):
        if self.async_reshuffle_in_progress():
            raise RuntimeError("Async reshuffle in progress, can't call reshuffle_async()")

        self._reshuffle_state["callback"] = callback
        self._reshuffle_prep(filter, dump_eval)

    def async_reshuffle_in_progress(self):
        return self._reshuffle_state["callback"] is not None

    def update(self):
        # If an async reshuffle is in progress
        if self.async_reshuffle_in_progress():
            self._reshuffle_do_chunk(self.async_reshuffle_count)
            if not self._reshuffle_state["to_process"]:
                self._reshuffle_finalize()

    def get_storylet(self, id):
        return self._all.get(id)
    
    def add_storylet(self, storylet):
        if storylet.id in self._all:
            raise ValueError(f"Storylet with id '{storylet.id}' already exists in the deck.")
        self._all[storylet.id] = storylet
    
    def draw(self):
        if self.async_reshuffle_in_progress():
            raise RuntimeError("Async reshuffle in progress, can't call draw()")

        self._current_draw += 1

        if not self._draw_pile:
            return None

        storylet = self._draw_pile.pop(0)
        if storylet.update_on_drawn:
            update_context(self.context, storylet.update_on_drawn)
        storylet.drawn(self._current_draw)
        return storylet

    def draw_hand(self, count, reshuffle_if_needed=False):
        storylets = []
        for _ in range(count):
            if len(self._draw_pile) == 0:
                if reshuffle_if_needed:
                    self.reshuffle()
                else:
                    break

            storylet = self.draw()
            if storylet is None:
                break

            storylets.append(storylet)

        return storylets
    
    def _reshuffle_prep(self, filter, dump_eval=None):
        # Empty the draw pile
        self._draw_pile = []
        self._reshuffle_state["dump_eval"] = dump_eval
        self._reshuffle_state["filter"] = filter
        # Temp map to hold lists by priority
        self._reshuffle_state["priority_map"] = {}
        # All the cards to walk
        self._reshuffle_state["to_process"] = list(self._all.values())

    def _reshuffle_do_chunk(self, count):
        number_to_do = min(count, len(self._reshuffle_state["to_process"]))

        while number_to_do > 0:
            number_to_do -= 1

            storylet = self._reshuffle_state["to_process"].pop(0)

            if not storylet.can_draw(self._current_draw):
                # Storylet fails the draw rules - skip
                continue

            # Apply filter, if available
            if self._reshuffle_state["filter"] is not None:
                if not self._reshuffle_state["filter"](storylet):
                    # Storylet fails the filter - skip
                    continue

            if not storylet.check_condition(self.context, self._reshuffle_state["dump_eval"]):
                # Storylet fails the condition - skip
                continue

            # Get the current priority for the storylet
            priority = storylet.calc_current_priority(
                self.context, self.use_specificity, self._reshuffle_state["dump_eval"]
            )

            if priority not in self._reshuffle_state["priority_map"]:
                # Does a priority list for this priority value exist in the temp map? If not, make it
                self._reshuffle_state["priority_map"][priority] = []

            # Add our card to this priority list
            self._reshuffle_state["priority_map"][priority].append(storylet)

    def _reshuffle_finalize(self):
        # Now sort all the resultant priorities
        sorted_priorities = sorted(self._reshuffle_state["priority_map"].keys(), reverse=True)

        # Shuffle each set of storylets that are the same priority. Then add them all to the master draw pile.
        # Result will be - higher priorities will be in the pile at the front, and we go down from there.
        for priority in sorted_priorities:
            bucket = self._reshuffle_state["priority_map"][priority]
            shuffle_array(bucket)
            self._draw_pile.extend(bucket)

        self._reshuffle_state["priority_map"] = None
        self._reshuffle_state["to_process"] = []
        self._reshuffle_state["filter"] = None
        callback = self._reshuffle_state["callback"]
        self._reshuffle_state["callback"] = None
        self._reshuffle_state["dump_eval"] = None
        if callback:
            callback()

    def dump_draw_pile(self):
        if self.async_reshuffle_in_progress():
            raise RuntimeError("Async reshuffle in progress, can't call dump_draw_pile()")

        return ",".join(storylet.id for storylet in self._draw_pile)
